### PR TITLE
Add reach only shuffle and aggregation phase.

### PR DESCRIPTION
**Pull request by ple13**

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)

**Review by SanjayVas — APPROVED** [review]
> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 4 unresolved discussions (waiting on @kungfucraig, @ple13, and @renjiezh)

___
*[`MODULE.bazel` line 328 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NwuZMjq4foq6GPefvV1:-NwuZMjq4foq6GPefvV2:b-n3bcur) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/MODULE.bazel#L328)):*
> ```Starlark
> )
> 
> archive_override(
> ```

nit: Add a DO_NOT_SUBMIT comment referencing the any-sketch PR so it's

1. Clear that this is intended to be changed before submitting/merging, and
2. Clear which pending PR this code comes from.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 200 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwua4YL2ZTM_6fiJZEn:-Nwua4YL2ZTM_6fiJZEo:b-bjm7mp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L200)):*
> ```cpp
>   RETURN_IF_ERROR(VerifySketchParameters(request.sketch_params()));
> 
>   // Verify that the ring modulus is a prime.
> ```

nit: Extract an `IsPrime` bare function. This way that function could be independently changed to e.g. use a more efficient implementation. It also makes the code self-documenting, making the comment unnecessary.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 477 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NwubdTyDqyiW4dneB8K:-NwubdTyDqyiW4dneB8L:bycqqmh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L477)):*
> ```cpp
>   // Count the non-zero registers.
>   int64_t non_empty_register_count = 0;
>   for (auto x : combined_sketch) {
> ```

nit: qualify `auto` as much as possible. In this case, I believe it can be `const auto`, since nothing will be mutating the copy of the integer.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 477 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwud6SR-GmPgAcBJ4ys:-Nwud6SR-GmPgAcBJ4yt:b-z9sc11) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L477)):*
> ```cpp
>   // Count the non-zero registers.
>   int64_t non_empty_register_count = 0;
>   for (auto x : combined_sketch) {
> ```

nit: prefer meaningful variable names. The common exception is loop indices, but this isn't an index.



_Code quote:_
```cpp
x
```

**Review by kungfucraig — APPROVED** [review]
> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 5 unresolved discussions (waiting on @ple13 and @renjiezh)

___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.h` line 28 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwuin8q2HXuQeKDpN9i:-NwuijEo5VMC1DmZ7wdU:blothjn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.h#L28)):*
> ```objc
>     CompleteShufflePhaseResponse;
> 
> absl::StatusOr<CompleteShufflePhaseResponse> CompleteShufflePhase(
> ```

nit: Can you add comments that Complete Shuffle Phase and CompleteAggregationPhase apply to Reach and Frequency measurements? You could consider renaming them in a follow-up PR.

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: 1 of 15 files reviewed, 2 unresolved discussions (waiting on @kungfucraig, @renjiezh, and @SanjayVas)

___
*[`MODULE.bazel` line 328 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NwuZMjq4foq6GPefvV1:-NwvrEvF6QfW1sPHqRAA:bvmmp4y) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/MODULE.bazel#L328)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: Add a DO_NOT_SUBMIT comment referencing the any-sketch PR so it's

1. Clear that this is intended to be changed before submitting/merging, and
2. Clear which pending PR this code comes from.
</blockquote></details>

Thanks. I've cut a release for any-sketch and it's ready now.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.h` line 28 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwuin8q2HXuQeKDpN9i:-NwvrNya3DBcZOsgPiuj:bo6m75r) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.h#L28)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

nit: Can you add comments that Complete Shuffle Phase and CompleteAggregationPhase apply to Reach and Frequency measurements? You could consider renaming them in a follow-up PR.
</blockquote></details>

I've changed the name of the functions.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 200 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwua4YL2ZTM_6fiJZEn:-NwxP0w3Bxdt7L6L94oD:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L200)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: Extract an `IsPrime` bare function. This way that function could be independently changed to e.g. use a more efficient implementation. It also makes the code self-documenting, making the comment unnecessary.
</blockquote></details>

Done.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 477 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NwubdTyDqyiW4dneB8K:-NwvyIQ5ArRJaRoFVS4_:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L477)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: qualify `auto` as much as possible. In this case, I believe it can be `const auto`, since nothing will be mutating the copy of the integer.
</blockquote></details>

Done.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 477 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwud6SR-GmPgAcBJ4ys:-NwvyHdE0mLONm91WpoT:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e5feede5ce0164dfa2b83696c7f336f3539d9921/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L477)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: prefer meaningful variable names. The common exception is loop indices, but this isn't an index.
</blockquote></details>

Done.

File: src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc [hunk 1]
```diff
@@ -126,7 +143,143 @@ absl::StatusOr<CompleteShufflePhaseResponse> CompleteShufflePhase(
...
+
+  // Sample a vector r of random values in [1, modulus).
+  ASSIGN_OR_RETURN(std::vector<uint32_t> r,
+                   prng->GenerateNonZeroUniformRandomRange(
+                       sketch_size, request.sketch_params().ring_modulus()));
+
+  // Transform share of non-zero registers to share of a non-zero random value.
+  for (int j = 0; j < sketch_size; j++) {
+    combined_sketch[j] =
+        (combined_sketch[j] * r[j]) % request.sketch_params().ring_modulus();
```
Comment (benjaminkreuter) [inline src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc:268]: Maybe it is also worth having a CHECK here?  As an alternative we can use 64 bit arithmetic here e.g. change the type of 'r' to 'std::vector<uint64_t>'.

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 15 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 5 unresolved discussions (waiting on @kungfucraig, @ple13, and @renjiezh)

___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 77 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwzw9cE5nq51_kZ3pYB:-Nwzw9cE5nq51_kZ3pYC:b-fip1aj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9a86ebc9b2941799be48c8ad1df91c2421e05a10/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L77)):*
> ```cpp
> // Checks if val is a prime.
> absl::StatusOr<bool> IsPrime(int val) {
>   if (val < 0) {
> ```

Just return false in this case so the return type can be `bool`. It's not incorrect to ask if a negative integer is prime, it's just that the answer is always "no".

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: 14 of 15 files reviewed, 5 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, @renjiezh, and @SanjayVas)

___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 76 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#gh-1589310297:-Nwzvmx82V7kyMLY20zc:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9a86ebc9b2941799be48c8ad1df91c2421e05a10/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L76)):*
<details><summary><i>Previously, benjaminkreuter wrote…</i></summary><blockquote>

This is an inefficient way to check primality.  It should be OK given how it is used, but we should be careful with this function (especially if we ever need a larger modulus); can you add a note about the O(sqrt(val)) running time to the comment above?
</blockquote></details>

Done.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 77 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-Nwzw9cE5nq51_kZ3pYB:-NwzyK9E7_rbROrn9k83:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9a86ebc9b2941799be48c8ad1df91c2421e05a10/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L77)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Just return false in this case so the return type can be `bool`. It's not incorrect to ask if a negative integer is prime, it's just that the answer is always "no".
</blockquote></details>

Done.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 246 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#gh-1589319471:-NwzyfmJ2pa5Jbkxj9mE:b-2v48wx) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9a86ebc9b2941799be48c8ad1df91c2421e05a10/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L246)):*
<details><summary><i>Previously, benjaminkreuter wrote…</i></summary><blockquote>

I think it is worth having a CHECK here.
</blockquote></details>

I've replaced it with a util function that handles the overflow.
___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 268 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#gh-1589331038:-NwzyrsR6mvfXHhhnKUN:b7e5bde) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9a86ebc9b2941799be48c8ad1df91c2421e05a10/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L268)):*
<details><summary><i>Previously, benjaminkreuter wrote…</i></summary><blockquote>

Maybe it is also worth having a CHECK here?  As an alternative we can use 64 bit arithmetic here e.g. change the type of 'r' to 'std::vector<uint64_t>'.
</blockquote></details>

Thanks Ben for catching this! I've updated the code.

**Review by kungfucraig — APPROVED** [review]
> Reviewed 14 of 15 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @renjiezh, and @SanjayVas)

**Review by SanjayVas — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter and @renjiezh)

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 14 of 15 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter and @renjiezh)

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter and @renjiezh)

**Review by renjiezh — CHANGES_REQUESTED** [review]
> Reviewed 7 of 15 files at r2, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter and @ple13)

___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 223 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NxxHtdeC9Xrpo3heuWF:-NxxHtdeC9Xrpo3heuWG:by3vw2r) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/45095fa29e681b1e0f2df8da396cd7b043a6bbc1/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L223)):*
> ```cpp
>   }
> 
>   const int sketch_size = request.sketch_params().register_count();
> ```

Do we need to check if `sketch_size` is 0?

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: 13 of 15 files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, @renjiezh, @SanjayVas, and @stevenwarejones)

___
*[`src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc` line 223 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600#-NxxHtdeC9Xrpo3heuWF:-NxxqwZWD7vixwjH4KHx:b-u6m92c) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/45095fa29e681b1e0f2df8da396cd7b043a6bbc1/src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc#L223)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

Do we need to check if `sketch_size` is 0?
</blockquote></details>

Yeah, I've added the check.

**Review by benjaminkreuter — APPROVED** [review]

**Review by renjiezh — APPROVED** [review]
> Reviewed 3 of 15 files at r2, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter)

**Review by ple13 — COMMENTED** [review]
> Reviewed 9 of 15 files at r2, 2 of 2 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter)

**Review by ple13 — COMMENTED** [review]
> Dismissed @benjaminkreuter from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1600)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ple13)

**Pull request merged**